numpyro/contrib/funsor/discrete.py

```diff
@@ -118,8 +118,7 @@ def _sample_posterior(
             values = [v.reshape((-1,) + prototype_shape[1:]) for v in values]
             data[root_name] = jnp.concatenate(values)
 
-    with substitute(data=data):
-        return model(*args, **kwargs)
+    return data
```

Comment:
I revise this a bit to use this function in Predictive (and avoid tracing the model 2 times)